contracts/storage/ServiceAgreementStorage.sol

```diff
@@ -0,0 +1,503 @@
+        bytes indexed keyword,
+        uint8 hashingFunctionId,
+        uint96 indexed identityId,
+        bytes indexed nodeId,
+        uint32 score
+    );
+    event ProofSubmitted(
+        address assetContract,
+        uint256 tokenId,
+        bytes indexed keyword,
+        uint8 hashingFunctionId,
+        uint96 indexed identityId,
+        bytes indexed nodeId
+    );
+
+
+    struct CommitSubmission {
+        uint96 identityId;
+        uint96 nextIdentity;
+        uint32 score;
+    }
+
+    struct ServiceAgreement {
+        uint256 startTime;
+        uint16 epochsNum;
+        uint128 epochLength;
+        uint96 tokenAmount;
+        uint8 scoringFunctionId;
+        uint8 proofWindowOffsetPerc;  // Perc == In % of the epoch
+        mapping(uint16 => bytes32) epochSubmissionHeads;  // epoch => headCommitId
+        mapping(uint16 => uint32) rewardedNodes;
+    }
+
+    Hub public hub;
+
+    // CommitId [keccak256(agreementId + epoch + identityId)] => CommitSubmission
+    mapping(bytes32 => CommitSubmission) commitSubmissions;
+
+    // hash(asset type contract + tokenId + key) -> ServiceAgreement
+    mapping(bytes32 => ServiceAgreement) public serviceAgreements;
+
+    constructor (address hubAddress) {
+        require(hubAddress != address(0));
+        hub = Hub(hubAddress);
+    }
+
+    modifier onlyAssetContracts() {
+        require (
+            hub.isAssetContract(msg.sender),
+            "Function can only be called by Asset Type Contracts"
+        );
+        _;
+    }
+
+    function createServiceAgreement(
+        address operationalWallet,
+        address assetContract,
+        uint256 tokenId,
+        bytes memory keyword,
+        uint8 hashingFunctionId,
+        uint16 epochsNum,
+        uint96 tokenAmount,
+        uint8 scoringFunctionId
+    )
+        public
+        onlyAssetContracts
+    {
+        IERC20 tokenContract = IERC20(hub.getContractAddress("Token"));
+        require(
+            tokenContract.allowance(operationalWallet, address(this)) >= tokenAmount,
+            "Sender allowance must be equal to or higher than chosen amount!"
+        );
+        require(
+            tokenContract.balanceOf(operationalWallet) >= tokenAmount,
+            "Sender balance must be equal to or higher than chosen amount!"
+        );
+
+        tokenContract.transferFrom(operationalWallet, address(this), tokenAmount);
+
+        ParametersStorage parametersStorage = ParametersStorage(hub.getContractAddress("ParametersStorage"));
+        uint8 minProofWindowOfssetPerc = parametersStorage.minProofWindowOffsetPerc();
+        uint8 maxProofWindowOffsetPerc = parametersStorage.maxProofWindowOffsetPerc();
+
+        bytes32 agreementId = _generateAgreementId(assetContract, tokenId, keyword, hashingFunctionId);
+
+        ServiceAgreement storage agreement = serviceAgreements[agreementId];
+        agreement.startTime = block.timestamp;
+        agreement.epochsNum = epochsNum;
```

Comment:
do we check if epochsNum > 0 somewhere?

Reply:
I guess I see in in asset contracts, but would be also useful that ServiceAgreementStorage takes care of this check too